### PR TITLE
 Consolidate PrintController code and cleanup some interop

### PR DESCRIPTION
**Pull request by hughbe**

### CODE REVIEW

File: src/Common/src/Interop/Windows/Kernel32/Interop.GlobalLock.cs [hunk 1]
```diff
@@ -10,9 +10,9 @@ internal partial class Interop
     internal partial class Kernel32
     {
         [DllImport(Libraries.Kernel32, PreserveSig = true)]
-        internal static extern IntPtr GlobalLock(IntPtr hMem);
+        internal static extern IntPtr GlobalLock(HandleRef hMem);
 
         [DllImport(Libraries.Kernel32, PreserveSig = true)]
-        internal static extern int GlobalUnlock(IntPtr hMem);
+        internal static extern int GlobalUnlock(HandleRef hMem);
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/Kernel32/Interop.GlobalLock.cs]: Our automated testing of System.Management is very limited.  Please don't replace these overloads that it's using.  You can just add the IntPtr overloads here.

Comment (hughbe) [reply src/Common/src/Interop/Windows/Kernel32/Interop.GlobalLock.cs]: Is there a difference between IntPtr and HandleRef marshalling? Just curious?

Comment (stephentoub) [reply src/Common/src/Interop/Windows/Kernel32/Interop.GlobalLock.cs]: Yes, HandleRef does additional stuff related to keeping alive a managed object during the P/Invoke.  It's kind of a precursor to SafeHandle.

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintController.Unix.cs [hunk 2]
```diff
@@ -1,59 +1,17 @@
...
-// included in all copies or substantial portions of the Software.
-// 
-// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
-// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
-// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
-// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
-// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
-// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
-// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
-//
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintController.Unix.cs]: Please put this back.

Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintController.Unix.cs]: Considering all the mono code is gone I thought it was safe to remove this but I wouldn’t want to upset any lawyers

**Comment by hughbe** [comment]
@stephentoub is this good to merge? :)

**Comment by danmosemsft** [comment]
@safern for that q

File: src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs [hunk 3]
```diff
@@ -15,63 +13,38 @@ namespace System.Drawing.Printing
...
-            _dc = document.PrinterSettings.CreateInformationContext(modeHandle);
+            _dc = document.PrinterSettings.CreateInformationContext(_modeHandle);
         }
 
         /// <summary>
         /// Implements StartEnd for generating print preview information.
         /// </summary>
         public override Graphics OnStartPage(PrintDocument document, PrintPageEventArgs e)
         {
-            Debug.Assert(_dc != null && _graphics == null, "PrintController methods called in the wrong order?");
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs]: Why remove this assert?

File: src/System.Drawing.Common/tests/Printing/PrintControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            var controller = new SubPrintController();
+            Assert.Throws<NullReferenceException>(() => controller.OnStartPrint(null, new PrintEventArgs()));
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        public void OnStartPrint_InvokeNullDocumentUnix_Nop()
+        {
+            var controller = new SubPrintController();
+            controller.OnStartPrint(null, new PrintEventArgs());
```
Comment (safern) [inline src/System.Drawing.Common/tests/Printing/PrintControllerTests.cs]: Should Unix and Windows have the same behavior?

File: src/System.Drawing.Common/tests/Printing/PrintControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+                controller.OnEndPrint(document, e);
+            }
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void OnStartPrint_InvokeNullDocumentWindows_ThrowsNullReferenceException()
+        {
+            var controller = new SubPrintController();
+            Assert.Throws<NullReferenceException>(() => controller.OnStartPrint(null, new PrintEventArgs()));
```
Comment (safern) [inline src/System.Drawing.Common/tests/Printing/PrintControllerTests.cs]: Instead of throwing a NRE could he throw an ArgumentNullException?

File: src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs (hunk 3 above)
Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs]: Easily triggered by calling without before calling `OnStartPrint`

Comment (safern) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs]: Should we instead throw an InvalidOperationException?

Comment (hughbe) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs]: We could - it would be breaking and require us to add a new exception message. What do you think?

Comment (safern) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs]: @stephentoub what do you think?

Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs]: What's the current release behavior in a situation where this assert would have been triggered?

Comment (safern) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Windows.cs]: From taking a look at the code I can see that we would Null Ref here: https://github.com/dotnet/corefx/pull/36736/files/558de14d62537ace81993796d54954b9c4b613f5#diff-368f90b5da0f52ee19e0acf2a4a54780R76 when calling: `_dc.Hdc`

**Comment by hughbe** [comment]
I rebased this PR. I think the decision we want is to fix the following tests:
```cs
[ConditionalTheory(Helpers.AnyInstalledPrinters, Helpers.IsDrawingSupported)]
public void OnStartPage_InvokeNullEventArgsPageSettings_ReturnsNull()
{
    using (var document = new PrintDocument())
    {
        var controller = new PreviewPrintController();
        controller.OnStartPrint(document, new PrintEventArgs());

        var printEventArgs = new PrintPageEventArgs(null, Rectangle.Empty, Rectangle.Empty, null);
        Assert.Throws<NullReferenceException>(() => controller.OnStartPage(document, printEventArgs));
    }
}

[Fact]
public void OnStartPage_PrintNotStarted_ThrowsNullReferenceException()
{
    using (var document = new PrintDocument())
    {
        var controller = new PreviewPrintController();
        var e = new PrintPageEventArgs(null, Rectangle.Empty, Rectangle.Empty, null);
        Assert.Throws<NullReferenceException>(() => controller.OnStartPage(document, e));
    }
}

[Fact]
[PlatformSpecific(TestPlatforms.Windows)]
public void OnEndPage_InvokeWithoutStartingWindows_ThrowsNullReferenceException()
{
    using (var document = new PrintDocument())
    {
        var controller = new PreviewPrintController();
        Assert.Throws<NullReferenceException>(() => controller.OnEndPage(document, new PrintPageEventArgs(null, Rectangle.Empty, Rectangle.Empty, null)));
        Assert.Throws<NullReferenceException>(() => controller.OnEndPage(null, null));
    }
}

[Theory]
[MemberData(nameof(PrintEventArgs_TestData))]
[PlatformSpecific(TestPlatforms.Windows)]
public void OnEndPrint_InvokeWithoutStartingWindows_Nop(PrintEventArgs e)
{
    using (var document = new PrintDocument())
    {
        var controller = new PreviewPrintController();
        Assert.Throws<NullReferenceException>(() => controller.OnEndPrint(document, e));
        Assert.Throws<NullReferenceException>(() => controller.OnEndPrint(null, e));
    }
}
```

**Comment by maryamariyan** [comment]
@hughbe is there gonna be more updates to this PR or is it ready for review?

**Comment by hughbe** [comment]
No more to add! Ready to be reviewed :)

**Comment by danmosemsft** [comment]
@maryamariyan perhaps you could finish the review here and get this merged?

File: src/System.Drawing.Common/tests/Printing/PreviewPrintControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void OnEndPage_InvokeWithoutStartingWindows_ThrowsNullReferenceException()
+        {
+            using (var document = new PrintDocument())
+            {
+                var controller = new PreviewPrintController();
+                Assert.Throws<NullReferenceException>(() => controller.OnEndPage(document, new PrintPageEventArgs(null, Rectangle.Empty, Rectangle.Empty, null)));
```
Comment (safern) [inline src/System.Drawing.Common/tests/Printing/PreviewPrintControllerTests.cs]: I think we should be consistent and throw NullReferenceException on both, Unix and Windows, any reason why you didn't do it?

File: src/System.Drawing.Common/tests/Printing/PreviewPrintControllerTests.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        [Theory]
+        [MemberData(nameof(PrintEventArgs_TestData))]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void OnEndPrint_InvokeWithoutStartingWindows_Nop(PrintEventArgs e)
+        {
+            using (var document = new PrintDocument())
+            {
+                var controller = new PreviewPrintController();
+                Assert.Throws<NullReferenceException>(() => controller.OnEndPrint(document, e));
+                Assert.Throws<NullReferenceException>(() => controller.OnEndPrint(null, e));
```
Comment (safern) [inline src/System.Drawing.Common/tests/Printing/PreviewPrintControllerTests.cs]: Ditto.

File: src/Common/src/Interop/Windows/Kernel32/Interop.GlobalLock.cs [hunk 8]
```diff
@@ -9,10 +9,16 @@ internal partial class Interop
...
-        [DllImport(Libraries.Kernel32, PreserveSig = true)]
-        internal static extern int GlobalUnlock(IntPtr hMem);
+        [DllImport(Libraries.Kernel32, ExactSpelling = true)]
+        public static extern IntPtr GlobalLock(HandleRef hMem);
+
+        [DllImport(Libraries.Kernel32, ExactSpelling = true)]
+        public static extern IntPtr GlobalUnlock(IntPtr hMem);
+
+        [DllImport(Libraries.Kernel32, ExactSpelling = true)]
+        public static extern IntPtr GlobalUnlock(HandleRef hMem);
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/Kernel32/Interop.GlobalLock.cs]: SetLastError = true is being removed from a bunch of these.  Does that mean we never check the error code after calling any of these?

File: src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs [hunk 9]
```diff
@@ -32,40 +32,31 @@
...
         public override void OnEndPage(PrintDocument document, PrintPageEventArgs e)
         {
         }
 
         public override void OnStartPrint(PrintDocument document, PrintEventArgs e)
         {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs]: Why is this necessary?  Internal callers of this won't pass null, and this shouldn't be called by external callers (even though technically it could be).  Why is it better to add code and cost rather than just letting it null ref when it's used in a way it shouldn't be used?

File: src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs [hunk 10]
```diff
@@ -74,31 +65,26 @@ public override void OnEndPrint(PrintDocument document, PrintEventArgs e)
...
         public override Graphics OnStartPage(PrintDocument document, PrintPageEventArgs e)
         {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+            if (e == null)
+            {
+                throw new ArgumentNullException(nameof(e));
+            }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs]: Same question for all of these additions.  They don't seem like the right trade-off to me.

File: src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs (hunk 9 above)
Comment (danmosemsft) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs]: +1, outside of public API this kind of check is probably only worthwhile in cases where it will not null ref til later (eg., on a constructor when the null ref occurs on a method later)

**Comment by hughbe** [comment]
Fortunately back from holiday tomorrow so can fix. Any feedback before then will be welcome :))

File: src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs (hunk 9 above)
Comment (safern) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PreviewPrintController.Unix.cs]: @hughbe planning on addressing these comments?
